pkg/aggregator/context_resolver.go

```diff
@@ -97,11 +128,56 @@ func (cr *ContextResolver) expireContexts(expireTimestamp float64) []ckey.Contex
 		}
 	}
 
+	cr.resolver.removeKeys(expiredContextKeys)
+
 	// Delete expired context keys
 	for _, expiredContextKey := range expiredContextKeys {
-		delete(cr.contextsByKey, expiredContextKey)
 		delete(cr.lastSeenByKey, expiredContextKey)
 	}
 
 	return expiredContextKeys
 }
+
+// countBasedContextResolver allows tracking and expiring contexts based on the number
+// of calls of `expireContexts`.
+type countBasedContextResolver struct {
+	resolver               *contextResolver
+	keyByContextCountIndex map[ckey.ContextKey]int64
+	contextCountIndex      int64
+	expireContextsCount    int64
+}
+
+func newCountBasedContextResolver(expireContextsCount int) *countBasedContextResolver {
+	return &countBasedContextResolver{
+		resolver:               newContextResolver(),
+		keyByContextCountIndex: make(map[ckey.ContextKey]int64),
+		contextCountIndex:      0,
+		expireContextsCount:    int64(expireContextsCount),
+	}
+}
+
+// trackContext returns the contextKey associated with the context of the metricSample and tracks that context
+func (cr *countBasedContextResolver) trackContext(metricSampleContext metrics.MetricSampleContext) ckey.ContextKey {
+	contextKey := cr.resolver.trackContext(metricSampleContext)
+	cr.keyByContextCountIndex[contextKey] = cr.contextCountIndex
+	return contextKey
+}
+
+func (cr *countBasedContextResolver) get(key ckey.ContextKey) (*Context, bool) {
+	return cr.resolver.get(key)
+}
+
+// expireContexts cleans up the contexts that haven't been tracked since `expirationCount`
+// call to `expireContexts` and returns the associated contextKeys
+func (cr *countBasedContextResolver) expireContexts() []ckey.ContextKey {
+	var keys []ckey.ContextKey
+	for key, index := range cr.keyByContextCountIndex {
+		if index <= cr.contextCountIndex-cr.expireContextsCount {
+			keys = append(keys, key)
+			delete(cr.keyByContextCountIndex, key)
+		}
+	}
+	cr.resolver.removeKeys(keys)
```

Comment:
Can this be moved inside the loop? I think this would save a slice allocation.

Reply:
I am not sure to understand what you mean.  `cr.resolver.removeKeys(keys)` should be called once and `keys` is returned by the function.

Reply:
Never mind, I missed the part where keys are returned.